### PR TITLE
Add deep links redaction

### PR DESCRIPTION
**Pull request by fathyb**
Ref: LIB-686

### CODE REVIEW

File: Analytics/Classes/Internal/SEGUtils.m [hunk 1]
```diff
@@ -34,4 +34,62 @@ + (id _Nullable)plistFromData:(NSData *_Nonnull)data
...
+            
+            NSInteger matches = [re replaceMatchesInString:str
+                                                   options:0
+                                                     range:NSMakeRange(0, str.length)
+                                              withTemplate:patterns[pattern]];
+            
+            if (matches > 0) {
+                SEGLog(@"%@ Redacted value from action: %@", self, str);
+                
+                return str;
```
Comment (nhi-nguyen) [inline Analytics/Classes/Internal/SEGUtils.m]: It looks like we're trying to match multiple patterns, but here we return immediately after the first match? Should this return statement be outside the `for` loop?

File: AnalyticsTests/AnalyticsTests.swift [hunk 2]
```diff
@@ -170,6 +170,29 @@ class AnalyticsTests: QuickSpec {
...
+      analytics.configuration.payloadFilters["(myapp://auth\\?token=)([^&]+)"] = "$1((redacted/my-auth))"
+      analytics.configuration.trackDeepLinks = true
+      analytics.open(URL(string: "myapp://auth?token=hastoberedacted&other=stuff")!, options: [:])
+      
+      
+      let event = testMiddleware.lastContext?.payload as? SEGTrackPayload
+      expect(event?.event) == "Deep Link Opened"
+      expect(event?.properties?["url"] as? String) == "myapp://auth?token=((redacted/my-auth))&other=stuff"
+    }
   }
```
Comment (nhi-nguyen) [inline AnalyticsTests/AnalyticsTests.swift]: can we have tests to make sure we're traversing json correctly, especially in the case of nested arrays/dicts?

File: Analytics/Classes/Internal/SEGUtils.m (hunk 1 above)
Comment (fathyb) [reply Analytics/Classes/Internal/SEGUtils.m]: Good catch! It's fixed

File: AnalyticsTests/AnalyticsTests.swift (hunk 2 above)
Comment (fathyb) [reply AnalyticsTests/AnalyticsTests.swift]: I added unit tests with nested structures

File: Analytics/Classes/SEGAnalyticsConfiguration.h [hunk 3]
```diff
@@ -140,4 +140,30 @@ typedef NSMutableURLRequest *_Nonnull (^SEGRequestFactory)(NSURL *_Nonnull);
...
+ * @endcode
+ *
+ * This will replace any matching occurences to a redacted version:
+ * @code
+ * "fb123456789://authorize#access_token=secretsecretsecretsecret&some=data"
+ * @endcode
+ *
+ * Becomes:
+ * @code
+ * "fb123456789://authorize#access_token=((redacted/fb-auth-token))"
```
Comment (f2prateek) [inline Analytics/Classes/SEGAnalyticsConfiguration.h]: given the example `fb123456789://authorize#access_token=secretsecretsecretsecret&some=data`, wouldn't the desired output be `fb123456789://authorize#access_token=((redacted/fb-auth-token))&some=data`? (i.e. `&some=data`  at the end.

**Comment by gerhard-segment** [comment]
Good point.

What is the risk of adding this ?

**Comment by f2prateek** [comment]
I just realized there is logging already - https://github.com/segmentio/analytics-ios/pull/798/files#diff-ecadcecb2e3287f62402c496069b01a2R85. This looks reasonable to me!

File: Analytics/Classes/SEGAnalyticsConfiguration.h (hunk 3 above)
Comment (fathyb) [reply Analytics/Classes/SEGAnalyticsConfiguration.h]: Facebook sends other private query parameters to the deep link that could be considered sensible, so we remove everything until we get a space (`[^ ]`).

Comment (f2prateek) [reply Analytics/Classes/SEGAnalyticsConfiguration.h]: Got it - sounds good 👍

**Pull request merged**